Fixed #4954: Check destination buffer bounds.

akallabeth opened the pull request:
Applies proper destination bound checks

mfleisz commented:
Did some performance testing because I wasn't sure about the impact of the checks in the write_pixel_ functions ... here are my results (times in ms):

```
Pre-Patch:
VS2010
100 runs took total 2233.29 (avg: 22.33)
100 runs took total 2213.39 (avg: 22.13)
100 runs took total 2236.61 (avg: 22.37)

VS2017
100 runs took total 2560.92 (avg: 25.61)
100 runs took total 2484.92 (avg: 24.85)
100 runs took total 2516.06 (avg: 25.16)

Post-Patch:
VS2010
100 runs took total 4225.69 (avg: 42.26)
100 runs took total 4234.21 (avg: 42.34)
100 runs took total 4236.19 (avg: 42.36)

VS2017
100 runs took total 3260.17 (avg: 32.60)
100 runs took total 3249.61 (avg: 32.50)
100 runs took total 3257.01 (avg: 32.57)
```

As can be seen this PR has quite an impact on performance. I also looked at the Assembly generated by VS2017 compiling interleaved.c (snippet from WriteFgBgImage24to24):

Before patch:
```
; 37   : 	else
; 38   : 	{
; 39   : 		DESTWRITEPIXEL(pbDest, xorPixel);

  00061	8b ca		 mov	 ecx, edx
$LN6@WriteFgBgI:

; 40   : 	}
; 41   : 
; 42   : 	DESTNEXTPIXEL(pbDest);
; 43   : 	cBits = cBits - 1;

  00063	8b 54 24 30	 mov	 edx, DWORD PTR cBits$[rsp]
  00067	45 88 13	 mov	 BYTE PTR [r11], r10b
  0006a	ff ca		 dec	 edx
  0006c	41 88 43 01	 mov	 BYTE PTR [r11+1], al
  00070	41 88 4b 02	 mov	 BYTE PTR [r11+2], cl
```

After patch:
```
; 38   : 	else
; 39   : 	{
; 40   : 		if (!DESTWRITEPIXEL(pbDest, pbDestEnd, xorPixel))
; 41   : 			return NULL;
; 42   : 	}
; 43   : 
; 44   : 	DESTNEXTPIXEL(pbDest);

  00046	4d 85 d2	 test	 r10, r10         <-- Buf null check
  00049	0f 84 da 00 00
	00		 je	 $LN66@WriteFgBgI
  0004f	4d 85 db	 test	 r11, r11         <-- Buf null check
  00052	0f 84 d1 00 00
	00		 je	 $LN66@WriteFgBgI
  00058	49 8b c3	 mov	 rax, r11
  0005b	49 2b c2	 sub	 rax, r10
  0005e	48 83 f8 03	 cmp	 rax, 3         <-- Buf size check
  00062	0f 8c c1 00 00
	00		 jl	 $LN66@WriteFgBgI

; 45   : 	cBits = cBits - 1;

  00068	44 8b 4c 24 58	 mov	 r9d, DWORD PTR cBits$[rsp]
  0006d	8b c1		 mov	 eax, ecx
  0006f	41 88 0a	 mov	 BYTE PTR [r10], cl
  00072	8b d1		 mov	 edx, ecx
  00074	c1 e8 08	 shr	 eax, 8
  00077	4d 8b c2	 mov	 r8, r10
  0007a	c1 ea 10	 shr	 edx, 16
  0007d	41 ff c9	 dec	 r9d
  00080	41 88 42 01	 mov	 BYTE PTR [r10+1], al
  00084	41 88 52 02	 mov	 BYTE PTR [r10+2], dl
```

As can be seen in the disassembly the nullptr checks and the length checks are done when writing each pixel with this patch applied.
The write_pixel_* functions should really just write pixels and buffer null/length checks should be done at a higher level.

hardening commented:
@mfleisz +1 for checks at the right place(tm)

akallabeth commented:
@mfleisz Ok, added a unit test with a profiler that encodes and decodes a random image. (with fuzzy image comparison, rounding error depending on color depth)
Both with `Release` flags and `address sanitizier` activated
Original `interleaved_decompress`
```
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - interleaved_compress   24bpp   |        500 |     0.0529s | 0.000106s |   9456
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - interleaved_decompress 24bpp   |        500 |     0.0307s | 0.000061s |  16302
[17:48:08:693] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - interleaved_compress   16bpp   |        500 |     0.0498s | 0.000100s |  10049
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - interleaved_decompress 16bpp   |        500 |     0.0276s | 0.000055s |  18133
[17:48:08:985] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - interleaved_compress   15bpp   |        500 |     0.0464s | 0.000093s |  10774
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - interleaved_decompress 15bpp   |        500 |     0.0260s | 0.000052s |  19212
[17:48:08:254] [15662:15662] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------

```

Refactored `interleaved_decompress`
```
[17:42:45:524] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:45:541] [13177:13177] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:42:45:541] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:45:541] [13177:13177] [INFO][com.freerdp.utils] - interleaved_compress   24bpp   |        500 |     0.0578s | 0.000116s |   8651
[17:42:45:541] [13177:13177] [INFO][com.freerdp.utils] - interleaved_decompress 24bpp   |        500 |     0.0346s | 0.000069s |  14433
[17:42:45:541] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - interleaved_compress   16bpp   |        500 |     0.0409s | 0.000082s |  12234
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - interleaved_decompress 16bpp   |        500 |     0.0245s | 0.000049s |  20423
[17:42:46:792] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - PROFILER NAME                  |      COUNT |       TOTAL |       AVG |    IPS
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - interleaved_compress   15bpp   |        500 |     0.0418s | 0.000084s |  11950
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - interleaved_decompress 15bpp   |        500 |     0.0249s | 0.000050s |  20086
[17:42:46:041] [13177:13177] [INFO][com.freerdp.utils] - -------------------------------+------------+-------------+-----------+-------

```

mfleisz commented:
Ran tests with the new code on VS2010 and VS2017 again:

Old code:
```
VC2010:
100 runs took total 2710.80 (avg: 27.11)
100 runs took total 2710.34 (avg: 27.10)
100 runs took total 2719.91 (avg: 27.20)

VC2017
100 runs took total 3052.95 (avg: 30.53)
100 runs took total 3082.49 (avg: 30.82)
100 runs took total 3056.62 (avg: 30.57)
```

New code:
```
VC2010:
100 runs took total 3737.04 (avg: 37.37)
100 runs took total 3748.09 (avg: 37.48)
100 runs took total 3741.23 (avg: 37.41)

VC2017:
100 runs took total 3863.13 (avg: 38.63)
100 runs took total 3885.83 (avg: 38.86)
100 runs took total 3787.79 (avg: 37.88)
```

From looking at the assembly code it looks like the compilers don't do any loop unrolling with the new code causing the performance hit.
